### ISSUE TITLE
Linking with Ssl causes segfault on multicore OCaml

### ISSUE THREAD

**Issue by talex5** [savonet/ocaml-ssl#76]
I've been investigating why Dream segfaults when run on multicore (see https://github.com/aantron/dream/issues/191).

I've managed to simplify the test-case to this (with `ssl` as its only dependency):

```ocaml
let foo = Ssl.file_descr_of_socket

let () =
  while true do
    Gc.full_major ()
  done
```

This segfaults immediately for me when run under `ocaml-variants.4.12.0+domains`.

I believe the cause is this bit of `ssl.ml`: https://github.com/savonet/ocaml-ssl/blob/b711cd4c155764bc3c0965ec9020f28575e186b3/src/ssl.ml#L191-L193

The C stub is https://github.com/savonet/ocaml-ssl/blob/b711cd4c155764bc3c0965ec9020f28575e186b3/src/ssl_stubs.c#L562-L565

However, `client_verify_callback` is not an OCaml value, it's a C function. This causes the GC to crash when trying to mark it.

**Comment by patricoferris** [comment]
I just bumped into this over at https://github.com/patricoferris/dream-multicore-repro

**Comment by anmonteiro** [comment]
This should have been fixed in #83 and released in 0.5.11.
